feat: add Exa search for related proposals and clickable proposal titles

JorikSchellekens opened the pull request:
## Summary

This PR implements Exa search functionality for finding related proposals and adds clickable links to original proposals.

## Features Added

### Backend Changes
- **Exa API Integration**: New service for semantic search across governance platforms
- **Related Proposals Endpoint**:  GET endpoint with authentication
- **Configuration**: Optional  environment variable
- **Error Handling**: Graceful degradation when Exa API key not configured

### Frontend Changes
- **RelatedProposals Component**: Displays semantically similar proposals from multiple platforms
- **Clickable Proposal Titles**: Direct links to original proposals on Snapshot
- **Loading States**: Proper loading and error handling UX
- **Responsive Design**: Clean UI matching existing theme

### Documentation
- **API Integrations Section**: Comprehensive documentation for all external APIs
- **Rate Limiting**: Documented Snapshot API limits (100 req/min without key)
- **Setup Instructions**: Clear guidance for Exa API configuration

## Platforms Supported
- Snapshot
- Arbitrum Forum  
- Aave Governance
- Compound
- Uniswap Governance
- MakerDAO Forum
- Tally Research
- Commonwealth

## Technical Implementation
- Follows existing authentication patterns
- Maintains backward compatibility
- Production-ready error handling
- Clean, accessible UI components
- Comprehensive test coverage ready

## Testing
- ✅ Backend compiles successfully
- ✅ Frontend linting passes
- ✅ No TypeScript errors
- ✅ Graceful degradation tested

JorikSchellekens committed:
feat: add Exa search for related proposals and clickable proposal titles

- Add Exa API integration for semantic search of related proposals
- Create /related-proposals endpoint with authentication
- Add RelatedProposals component displaying search results
- Make proposal titles clickable links to original Snapshot proposals
- Update configuration for optional Exa API key
- Document API integrations and rate limiting in README
- Support graceful degradation when Exa API key not configured

JorikSchellekens committed:
fix: remove needless borrow in Exa service

- Fix clippy warning about needless_borrows_for_generic_args

JorikSchellekens committed:
Add fmt and clippy rules to create-pr

JorikSchellekens commented:
fixes #101 and #86

yehia67 commented on crates/agent/src/services/exa.rs:
```diff
@@ -0,0 +1,197 @@
+}
+
+/// Extract domain name from URL for display purposes
+fn extract_domain(url: &str) -> String {
+    if let Ok(parsed_url) = url::Url::parse(url) {
+        if let Some(domain) = parsed_url.domain() {
+            return domain.to_string();
+        }
+    }
+    "Unknown".to_string()
+}
```
```suggestion
fn extract_domain(url: &str) -> String {
    url::Url::parse(url)
        .ok()
        .and_then(|u| u.domain().map(|d| d.to_string()))
        .unwrap_or_else(|| "Unknown".into())
}
```

varex83 replied:
Even better: `unwrap_or` :D

JorikSchellekens replied:
resolved

varex83 commented on crates/agent/src/api/handlers.rs:
```diff
@@ -82,3 +85,54 @@ pub struct AnalyzeResponse {
+    pub related_proposals: Vec<RelatedProposal>,
+    /// The query that was used for the search
+    pub query: String,
+}
```
better to put it into models directory

JorikSchellekens replied:
resolved

varex83 commented on crates/agent/src/services/exa.rs:
```diff
@@ -0,0 +1,197 @@
+                .await
+                .unwrap_or_else(|_| "Unknown error".to_string());
+            error!("Exa API error {}: {}", status, error_text);
+            return Err(anyhow!(
```
We don't use anyhow results, so better to refactor it to use our error type

JorikSchellekens replied:
Can you put this in the cursor and claude rules

JorikSchellekens replied:
resolved

varex83 commented on crates/agent/src/api/handlers.rs:
```diff
@@ -82,3 +85,54 @@ pub struct AnalyzeResponse {
+        .config
+        .exa_api_key
+        .as_ref()
+        .ok_or_else(|| ApiError::internal_error("Exa API key not configured"))?;
```
```suggestion
        .ok_or(ApiError::internal_error("Exa API key not configured"))?;
```

JorikSchellekens committed:
fix: resolve PR review comments

- Move RelatedProposalsQuery and RelatedProposalsResponse structs to models directory
- Refactor Exa service to use project error types instead of anyhow
- Simplify extract_domain function using unwrap_or pattern
- Simplify ok_or_else to ok_or in handlers.rs
- Fix all clippy warnings and ensure code quality
